### PR TITLE
parts: refactor to remove stored dictionary data

### PR DESCRIPTION
**Pull request by cmatsuoka**
Refactor the parts definition to unmarshal parts specification data
instead of storing it. Validation should be done before populating the
spec object instead of relying on jsonschema validation.

This PR conflicts with PR #6.

Signed-off-by: Claudio Matsuoka <claudio.matsuoka@canonical.com>

- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

-----

### CODE REVIEW

File: craft_parts/unmarshal.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        return value
+
+    def pop_boolean(self, key: str, default: bool = False) -> bool:
+        """Pop an item and validates it as a boolean."""
+        value = self._data.pop(key, default)
+        if not isinstance(value, bool):
+            raise ValueError(f"{key!r} must be a boolean")
+        return value
+
+    def pop_list_str(self, key: str, default: List[str] = None) -> List[str]:
```
Comment (cjp256) [inline craft_parts/unmarshal.py:59]:
```suggestion
    def pop_list_str(self, key: str, default: Optional[List[str]] = None) -> List[str]:
```

File: craft_parts/unmarshal.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        self._data = data
+
+    def pop_string(self, key: str, default: str = "") -> str:
+        """Pop an item and validates it as a string."""
+        value = self._data.pop(key, default)
+        if not isinstance(value, str):
+            raise ValueError(f"{key!r} must be a string")
+        return value
+
+    def pop_optional_string(self, key: str, default: str = None) -> Optional[str]:
```
Comment (cjp256) [inline craft_parts/unmarshal.py:38]:
```suggestion
    def pop_optional_string(self, key: str, default: Optional[str] = None) -> Optional[str]:
```

File: tests/unit/test_unmarshal.py [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+    "data,value,error",
+    [
+        [{}, "", None],
+        [{"test": ""}, "", None],
+        [{"test": "foo"}, "foo", None],
+        [{"test": False}, "", "'test' must be a string"],
+    ],
+)
+def test_unmarshal_string(data, value, error):
+    udata = unmarshal.DataUnmarshaler(data)
```
Comment (cjp256) [inline tests/unit/test_unmarshal.py:32]: I'd suggest breaking out the error scenarios into its own test(s) to reduce test complexity.

File: craft_parts/errors.py [hunk 4]
```diff
@@ -81,3 +81,18 @@ def __init__(self, arch_name: str):
...
         super().__init__(brief=brief, resolution=resolution)
+
+
+class PartSpecificationError(PartsError):
+    """Parts were not correctly specified.
+
+    :param message: The error message.
+    """
+
+    def __init__(self, part_name: str, message: str):
```
Comment (cjp256) [inline craft_parts/errors.py:92]: I have a strong personal preference for kwargs when the args are not intuitive :D 
```suggestion
    def __init__(self, *, part_name: str, message: str):
```

File: craft_parts/parts.py [hunk 5]
```diff
@@ -16,15 +16,137 @@
...
+
+        :raise ValueError: If data validation fails.
+        """
+        if not isinstance(data, dict):
+            raise ValueError("part data is not a dictionary")
+
+        udata = unmarshal.DataUnmarshaler(data)
+
+        return cls(
+            plugin=udata.pop_optional_string("plugin"),
```
Comment (cjp256) [inline craft_parts/parts.py:76]: I will say that seeing pop in unmarshal is a bit odd.  From a library level, the passed in dict is going to be deepcopy()'d before craft-parts modifies it, right?

Comment (cjp256) [reply craft_parts/parts.py:76]: And the popping is just to ensure there are no extra fields after Parts and Plugin?

If the plugin properties were stripped first (i.e. `plugin: make`  filters out `make-*`, then unmarshal can sanity check by doing its own `data.copy()` and check for leftover fields before returning.

File: tests/unit/test_unmarshal.py (hunk 3 above)
Comment (cmatsuoka) [reply tests/unit/test_unmarshal.py:32]: Reorganize tests to make them easier to read.

File: craft_parts/parts.py (hunk 5 above)
Comment (cmatsuoka) [reply craft_parts/parts.py:76]: Changing to make data consumption optional (by passing `consume=True` in `DataUnmarshaler` constructor) and rename `pop_*` methods to `get_*` to make the utility more generic, implementing the regular non-consuming case as the default. Then use the special consuming case in part specification unmarshal to detect if fields left without changing caller data, raising an error if unexpected fields found. Reorganize tests.

**Comment by cmatsuoka** [comment]
Superseded by PR #9.

**Pull request closed without merging**